format/moc3/moc3.go

```diff
@@ -0,0 +1,1766 @@
+		keyformColorSourcesBeginIndices int64
+	}
+
+	rotationDeformersV4_2 struct {
+		keyformColorSourcesBeginIndices int64
+	}
+
+	artMeshesV4_2 struct {
+		keyformColorSourcesBeginIndices int64
+	}
+
+	keyformMultiplyColors struct {
+		r int64
+		g int64
+		b int64
+	}
+
+	keyformScreenColors struct {
+		r int64
+		g int64
+		b int64
+	}
+
+	parametersV4_2 struct {
+		parameterTypes                                int64
+		blendShapeParameterBindingSourcesBeginIndices int64
+		blendShapeParameterBindingSourcesCounts       int64
+	}
+
+	blendShapeParameterBindings struct {
+		keysSourcesBeginIndices int64
+		keysSourcesCounts       int64
+		baseKeyIndices          int64
+	}
+
+	blendShapeKeyformBindings struct {
+		parameterBindingSourcesIndices               int64
+		blendShapeConstraintIndexSourcesBeginIndices int64
+		blendShapeConstraintIndexSourcesCounts       int64
+		keyformSourcesBlendShapeIndices              int64
+		keyformSourcesBlendShapeCounts               int64
+	}
+
+	blendShapesWarpDeformers struct {
+		targetIndices                               int64
+		blendShapeKeyformBindingSourcesBeginIndices int64
+		blendShapeKeyformBindingSourcesCounts       int64
+	}
+
+	blendShapesArtMeshes struct {
+		targetIndices                               int64
+		blendShapeKeyformBindingSourcesBeginIndices int64
+		blendShapeKeyformBindingSourcesCounts       int64
+	}
+
+	blendShapeConstraintIndices struct {
+		blendShapeConstraintSourcesIndices int64
+	}
+
+	blendShapeConstraints struct {
+		parameterIndices                             int64
+		blendShapeConstraintValueSourcesBeginIndices int64
+		blendShapeConstraintValueSourcesCounts       int64
+	}
+
+	blendShapeConstraintValues struct {
+		keys    int64
+		weights int64
+	}
+
+	blendShapesParts struct {
+		targetIndices                               int64
+		blendShapeKeyformBindingSourcesBeginIndices int64
+		blendShapeKeyformBindingSourcesCounts       int64
+	}
+
+	blendShapesRotationDeformers struct {
+		targetIndices                               int64
+		blendShapeKeyformBindingSourcesBeginIndices int64
+		blendShapeKeyformBindingSourcesCounts       int64
+	}
+
+	blendShapesGlue struct {
+		targetIndices                               int64
+		blendShapeKeyformBindingSourcesBeginIndices int64
+		blendShapeKeyformBindingSourcesCounts       int64
+	}
+}
+
+func decodeMOC3(d *decode.D) any {
+	d.FieldUTF8("magic", 4, d.StrAssert("MOC3"))
+	version := d.FieldU8("version", moc3VersionNames)
+	isBigEndian := d.FieldBoolFn("is_big_endian", func(d *decode.D) bool { return d.U8() != 0 })
+
+	if !isBigEndian {
+		d.Endian = decode.LittleEndian
+	}
+	d.FieldRawLen("unused", 58*8)
+
+	var sectionOffsets sectionOffsetTable
+	d.FramedFn(0x280*8, func(d *decode.D) {
+		d.FieldStruct("section_offsets", func(d *decode.D) {
+			sectionOffsets.countInfo = int64(d.FieldU32("count_info"))
+			sectionOffsets.canvasInfo = int64(d.FieldU32("canvas_info"))
+
+			d.FieldStruct("parts", func(d *decode.D) {
```

Comment:
This code makes me think that maybe the decode API should have something similar to `Read` in `encoding/binary` hmm

Reply:
`FieldUnmarshalStruct(field string, out any)`?

Reply:
Yeap something like that. Auto camel_case names and maybe use struct tags to support some casting. Something for the future, no need for this PR.